### DISCUSSION TITLE
How to structure a complex query with select and join tables

### DISCUSSION THREAD

**Discussion by cspecter** [supabase/supabase#1393] [Questions] [answered]
I am migrating a DB over from Firebase and I'm trying to figure out how to structure a fairly complex query that combines several tables via  join tables. 

In a simplified way my data looks like this. I have a list of Products. Each product had several tags, which I connected via a join table. Users can save products to a wishlist, again through a join table. Users can also follow other users. It looks sort of like this. 

```

// Base types
type product_tags = {
    id: string
    name: string
}

type product = {
    id: string
    name: string
    price: number
    description: string
    tags: product_tags[]
}

type user = {
    id: string
    username: string
    location: string
    profilePicture: string
    wishlist: product[]
    wishlistCount: number
    follows: user[]
    collorCount: number
}

// Join tables

type product_product_tags = {
    id: number
    product_id: string
    product_tag_id: string
}

type wishlist = {
    id: number
    product_id: string
    user_id: string
}

type relationships = {
    id: number
    follower_id: string
    followee_id: string
}

```

I would like to make one call for the user and get the following data:

```

// In graphQL nomenclature I want something structured like this
{
  id
  username
  location
  profilePicture
  wishlist {
    id
    name
    price
    description
    tags {
      id
      name
    }
  }
  wishlistCount
  follows {
     id
     username
     profilePicture
  }
  followCount
}

// With a returned object that looks like this

const data = {
  id: 'ABCD1234...',
  username: 'JohnSmith',
  location: 'Somwhere, USA',
  profilePicture: 'http://...',
  wishlist: [
     {
         id: 'POIU9876..',
         name: 'Product Name',
         price: 99.99,
         description ' ... ', 
         tags: [
            { id: 1, name: 'Tag1' },
            { id: 2, name: 'Tag2' },
            ...
         ]
     }
  ],
  wishlistCount; 50,
  follows: [
    {
       id: 'ABCD1234...',
       username: 'JaneSmith',
        profilePicture: 'http://...',
    },
    ...
  ],
  followCount: 10
}

```

Using the Supabase-js API how would I structure the select string?
```

const { data, error } = await supabase
  .from('users')
  .select(`
    id,
    username,
    location,
    profilePicture,
    wishlist: ??????????
  `)

```

**Comment by steve-chavez** [comment, marked as answer]
Hey @cspecter,

You should be able to do something like this:

```js
const { data, error } = await supabase
  .from('users')
  .select(`
    id,
    username,
    location,
    profilePicture,
    wishlistProducts:products(
      id,
      name,
      price,
      description,
      tags(
        id,
        name
      )
    ),
    wishlistCount:wishlist(count),
    follows(
      id,
      username,
      profilePicture
    ),
    followCount:follows(count)
  `)
```

This is assuming all the foreign keys between the tables are setup correctly.

**Reply by fergusmeiklejohn** [reply]
I'm coming from firestore and very new to SQL, so can I ask please:
Is this very useful query relying on their being arrays of foreign keys? => `follows: user[]` can be queried to return an array of user objects: `{  id, username, profilePicture }`, by iterating over the list of foreign keys, each of which is a user table primary key?
This would solve lots of problems I also have migrating from Firestore to Supabase, but I've read that postgres doesn't support foreign keys in arrays?? Is this still true and does it really matter?

**Reply by steve-chavez** [reply]
> but I've read that postgres doesn't support foreign keys in arrays?? 

Yes, this is true, but you can use regular join tables to support many to many relationships. For more details, check this previous discussion https://github.com/supabase/supabase/discussions/535#discussioncomment-285696.

**Reply by fergusmeiklejohn** [reply]
Thanks Steve, I wonder which option people coming from Firestore will prefer? The array method feels more familiar from a NoSQL point of view. Then we'd use a cloud function to delete orphan ids when the parent is deleted

**Reply by cspecter** [reply]
Thanks Steve,

You suggestion worked, though we are now exploring a Postgraphile solution. We had rolled a custom Apollo Server to interface with Firebase and automatically generating a GraphQL schema seems attractive. But handling join tables is ugly at best. The Supabase solution seems cleaner, though we would have to write our own logic to dynamically create the select string. We ultimately want a GraphQL server either way.

**Reply by nextdev1111** [reply]
Thanks bro

**Reply by jtkeyva** [reply]
@steve-chavez thanks! can that query live on the server and be called remotely?

**Reply by curiouscod3** [reply]
This is a very useful tip due to the lack of a detail/complex example documentation on the official site. I'm also migrating from Firebase to Supabase but has tons of issues ! kind of painful

**Comment by enzo-matrix01** [comment]
Hey, just want to add this detail. I have a profiles table that owns many reviews.
On reviews, I defined both author (review author) as a profile, and user_id (person being reviewed) as a profile, so I had to update my query as per this error:
```
{
  message: "Could not embed because more than one relationship was found for 'profiles' and 'reviews'",
  details: [
    {
      relationship: 'reviews_author_fkey[id][author]',
      embedding: 'profiles with reviews',
      cardinality: 'one-to-many'
    },
    {
      relationship: 'reviews_user_id_fkey[id][user_id]',
      embedding: 'profiles with reviews',
      cardinality: 'one-to-many'
    }
  ],
  hint: "Try changing 'reviews' to one of the following: 'reviews!reviews_author_fkey', 'reviews!reviews_user_id_fkey'. Find the desired relationship in the 'details' key."
}
```

**Reply by maxmdev** [reply]
Thank you, guys! Really helped a lot!
